tests/Feature/Livewire/UsernameInputTest.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace Tests\Feature\Livewire;
+
+use App\Actions\GenerateUsername;
+use App\Http\Livewire\UsernameInput;
+use App\Models\User;
+use Illuminate\Support\Str;
+use Livewire\Livewire;
+use Tests\TestCase;
+
+class UsernameInputTest extends TestCase
+{
+    public function test_component_can_render(): void
+    {
+        $component = Livewire::test(UsernameInput::class);
+
+        $component->assertStatus(200);
+    }
+
+    public function test_has_data_passed_correctly(): void
+    {
+        $value = 'just-a-test';
+        $placeholder = 'placeholder';
+        $label = 'label';
+        $required = true;
+
+        Livewire::test(
+            UsernameInput::class,
+            [
+                'value' => $value,
+                'placeholder' => $placeholder,
+                'label' => $label,
+                'required' => $required,
+            ]
+        )
+            ->assertSet('value', $value)
+            ->assertSet('placeholder', $placeholder)
+            ->assertSet('label', $label)
+            ->assertSet('required', $required);
+    }
+
+    /**
+     * @dataProvider validationDataProvider
+     */
+    public function test_value_has_proper_validation(string $fieldKey, array $inputData): void
+    {
+        User::factory()->create(
+            [
+                'email' => 'test@test.com',
+                'username' => 'test',
+            ]
+        );
+        Livewire::test(UsernameInput::class)
+            ->set($fieldKey, $inputData[$fieldKey])
+            ->assertHasErrors([$fieldKey]);
+    }
+
+    public function test_value_follows_specific_rule(): void
+    {
+        Livewire::test(UsernameInput::class)
+            ->set('value', 'Another bad username')
+            ->assertSee(
+                'The value must be valid. Only lowercase ASCII characters are allowed. Hyphens can be used. Whitespace, underscores, and multiple hyphens are not permitted. The username cannot start or end with a hyphen.'
```

Comment:
Should I test for the whole message? I saw that others do indeed test for this error messages, especially if is a custom rule, but not sure if is ok.

Reply:
I don't think it's necessary, I wouldn't do it. But you are right, some people are doing it. As long as there is an error, we are good.

Reply:
But the way, I'm not an expert at this :) Just sharing my thoughts. Thanks for PRs, I like this feeling like you are a part of something that is not linked to money in any way. Open-source collaboration feels great

Reply:
I will remove this test case where I check for the specific message, as it doesn't really matter that much to know what message it returned.

Also I feel the same as you regarding open source collaboration, another benefit is that you learn a lot of new things by being exposed to different opinions.

Btw thanks for the code reviews I appreciate it.

Reply:
@SerhiiCho @brendt Whenever you have time have a look at the POC that I did, the tests seem to pass, I did have a small problem with `min:1` test case, I will have to check what is the problem with that one for now I changed the rule to have a minimum of 2.

This is a POC, just wanna explore some ideas.